utils/near-rate-limiter/src/framed_read.rs

```diff
@@ -0,0 +1,380 @@
+        self.total_sizeof_messages_in_progress.fetch_add(msg_size, Ordering::SeqCst);
+    }
+
+    // Decrease limits by size of the message and notify ThrottledFramedReader to try to
+    // read again
+    pub fn remove_msg(&mut self, msg_size: usize) {
+        self.num_messages_in_progress.fetch_sub(1, Ordering::SeqCst);
+        self.total_sizeof_messages_in_progress.fetch_sub(msg_size, Ordering::SeqCst);
+
+        // Notify throttled framed reader
+        self.tx.send(()).expect("sending msg to tx failed");
+    }
+}
+
+impl<T, D> ThrottledFrameRead<T, D>
+where
+    T: AsyncRead,
+    D: Decoder,
+{
+    /// Creates a new `NearFramedRead` with the given `decoder`.
+    pub fn new(
+        inner: T,
+        decoder: D,
+        rate_limiter: RateLimiterHelper,
+        receiver: UnboundedReceiver<()>,
+    ) -> ThrottledFrameRead<T, D> {
+        ThrottledFrameRead {
+            inner: FramedImpl {
+                inner,
+                codec: decoder,
+                state: Default::default(),
+                rate_limiter,
+                receiver,
+            },
+        }
+    }
+}
+
+impl<T, D> ThrottledFrameRead<T, D> {
+    pub fn get_ref(&self) -> &T {
+        &self.inner.inner
+    }
+
+    pub fn into_inner(self) -> T {
+        self.inner.inner
+    }
+
+    pub fn decoder(&self) -> &D {
+        &self.inner.codec
+    }
+
+    pub fn read_buffer(&self) -> &BytesMut {
+        &self.inner.state.buffer
+    }
+}
+
+// This impl just defers to the underlying FramedImpl
+impl<T, D> Stream for ThrottledFrameRead<T, D>
+where
+    T: AsyncRead,
+    D: Decoder,
+{
+    type Item = Result<D::Item, D::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().inner.poll_next(cx)
+    }
+}
+
+pub(crate) struct ReadFrame {
+    pub(crate) eof: bool,
+    pub(crate) is_readable: bool,
+    pub(crate) buffer: BytesMut,
+}
+
+impl Default for ReadFrame {
+    fn default() -> Self {
+        Self { eof: false, is_readable: false, buffer: BytesMut::with_capacity(INITIAL_CAPACITY) }
+    }
+}
+
+impl<T, U, R> Stream for FramedImpl<T, U, R>
+where
+    T: AsyncRead,
+    U: Decoder,
+    R: BorrowMut<ReadFrame>,
+{
+    type Item = Result<U::Item, U::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut pinned = self.project();
+        let state: &mut ReadFrame = pinned.state.borrow_mut();
+        loop {
+            while let task::Poll::Ready(_) = pinned.receiver.poll_recv(cx) {
+                // pop all elements from queue, to prevent memory leak
+            }
+            while !pinned.rate_limiter.is_ready() {
+                // This will cause us to subscribe to notifier when something gets pushed to
+                // `pinned.receiver`. If there is an element in the queue, we will check again.
+                ready!(pinned.receiver.poll_recv(cx));
+            }
```

Comment:
Hm, so basically what we are doing is simulating level triggering (there's some message in channel) via edge triggerring (we receieved a new message). Intuitively, it seems like there should be a more direct synchronization primitive for this case. The right way place to shop for them is the `tokio::sync` module 

https://docs.rs/tokio/1.13.0/tokio/sync/index.html#structs

Looking at the list, it seems that `Notify` is what we need here: 

https://docs.rs/tokio/1.13.0/tokio/sync/struct.Notify.html

Reply:
I tried to use `Notify`, but I couldn't get it to work. I'm having trouble writing code, which pools from it using `notify2.notified()`

Consider the following code:
```
    let notify = Arc::new(Notify::new());

    tokio::spawn(async move {
        notify2.notified().await;
        println!("received notification");
    });
```
 notify2.notified() returns `Notified<'_>`. In order to pool on a future, I need to pin it.
However, `Notified` keep a reference to `Notify`, that makes things more complicated to pool.

Reply:
Ok, I added TODO.

I see that both `Notify` and `Semaphore` have a similar issue.

Reply:
Aha, indeed missed that. Yeah, I believe there aren't super-nice solutions here, so probably a comment is all there is needed. 

For semaphore, tokio hash https://docs.rs/tokio-util/0.6.9/tokio_util/sync/struct.PollSemaphore.html specifically for this case. So, I guess a long-term soln here is to send a PR to tokio-utils adding `PollNotify`?

Reply:
Ok, I started using `PoolSemaphore`, the code is much cleaner, thanks.

Yes, adding `PollNotify` PR seems like a good idea.

Reply:
Anyway, I'm using now the `PoolSemaphore` as I would `PollNotify`. So it's not an urgent thing.